### PR TITLE
move code

### PR DESCRIPTION
**Pull request by jquense**
Moving code over from fb/react

I’m not entirely sure how to handle licensing in this case..

### CODE REVIEW

**Comment by jquense** [comment]
There where a few (a lot) of internal api usage, most was easy enough to fix

- `dom-helpers` instead of CSSCore, we use it on react-bootstrap so probably a safe swap.
- Remove the transition event code branch entirely (it was already deprecated) and depend on timeouts. we can always use the dom-helpers version if we want to add it back
- Do some really hacky stuff with Children to avoid needing `flattenChildren`

**Comment by jquense** [comment]
cc @gaearon

**Comment by gaearon** [comment]
cc @zpao, how should we handle LICENSE/PATENTS files and copyright headers?

**Comment by taion** [comment]
Wasn't there a question a bit ago about support for raw numbers between CSSCore and dom-helpers? Or is that not applicable here?

**Comment by gaearon** [comment]
>Remove the transition event code branch entirely (it was already deprecated) and depend on timeouts. we can always use the dom-helpers version if we want to add it back

Can we keep it for one version? So that initial release is functionally identical to that provided by React 15. Then we can cut 2.x that drops it.

**Comment by jquense** [comment]
> Can we keep it for one version? So that initial release is functionally identical to that provided by React 15. Then we can cut 2.x that drops it.

sure, no problem

**Comment by jquense** [comment]
> Wasn't there a question a bit ago about support for raw numbers between CSSCore and dom-helpers? Or is that not applicable here?

yes but in case its just using the className utils, not styles

File: src/utils/ChildMapping.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ */
+export function getChildMapping(children) {
+  if (!children) {
+    return children;
+  }
+  let result = {};
+  Children.map(children, child => child)
+    .forEach((child) => {
+      result[child.key] = child;
+    });
```
Comment (gaearon) [inline src/utils/ChildMapping.js]: Does something like `Children.toArray(children)` help? I don't actually know this part of the code well.

Comment (jquense) [inline src/utils/ChildMapping.js]: I don't think so, or at least not fully. 

from what I gather, `flattenChildren` will return  object where the keys are the react key/data-id and values the corresponding children. my code above is attempting to emulate it by running it through `Children.map` which flattens and  adds keys to unkeyed elements. The forEach turns it into a hash.

The double iteration isn't great, but also we lose the extra debug info that `flattenChildren` gave, pointing out duplicates keys.

**Comment by gaearon** [comment]
Let's do it 👍

**Comment by jquense** [comment]
ok last bit added 👍

**Comment by jquense** [comment]
ok released! If all want to get the word out for folks to try it out, to make sure I didn't break anything in the move.

**Comment by gaearon** [comment]
Do you mind forking the docs too?

**Comment by gaearon** [comment]
Docs have a separate license: https://github.com/facebook/react/blob/master/LICENSE-docs

**Comment by jquense** [comment]
yes eventually, but a bit short on time right now! For now though, i don't think linking to the original docs places should be licensing issue right?

**Comment by gaearon** [comment]
Sure it's fine. I just mean in case you want to make API changes or anything.

Also the docs refer to `React.addons` so beginners will likely not realize they need to use a different package.

**Comment by jquense** [comment]
yeah its not a great solution in the medium to long term. We'll get some proper docs up on a gh-pages soon as I have some time to fork and extract the current ones

**Comment by gaearon** [comment]
I think it's also fine just to copy & paste this on README and leave it on GitHub.
You'll miss out on our custom line highlighting but otherwise it's okay.

**Comment by jquense** [comment]
ha true, I forgot it's all markdown :P

**Pull request merged**